Add LEP - Support K8s Cluster Autoscaler (#3831)
longhorn/longhorn: enhancements/20220408-support-kubernetes-ca.md

# Support Kubernetes Cluster Autoscaler

Longhorn should support Kubernetes Cluster Autoscaler.

## Summary

Currently, Longhorn pods are [blocking CA from removing a node](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/FAQ.md#what-types-of-pods-can-prevent-ca-from-removing-a-node). This proposes to introduce a new global setting `kubernetes-cluster-autoscaler-enabled` that will annotate Longhorn components and also add logic for instance-manager PodDisruptionBudget management.

### Related Issues

https://github.com/longhorn/longhorn/issues/2203

## Motivation

### Goals

- Longhorn should block CA from scaling down if a node met ANY condition:

>     - Volume attached

Reviewer: ```suggestion
    - Any volume attached
```

Author: Fixed.

>     - Contain backing image pod

Author: Fixed.

>     - Contain share manager pod

Author: Fixed.

- Longhorn should not block CA from scaling down if a node met ALL conditions:

>     - Volume detached and can find another schedulable node with volume replica and replica IM PDB.

Reviewer: ```suggestion
    - All volume detached, replicas of each detailed volume should be replicated to other schedulable nodes protected by replica IM PDB.
```

Author: Hmm..the suggested change looks like we will rebuild the replicas to another node. 

This is a condition statement to ensure the last replica of the detached volume is protected and we do not rebuild replicas to another node.

>     - Not contain a backing image pod

Reviewer: ```suggestion
    - Not contain a backing image manager pod
```

Author: Fixed.

- CA blocked by share manager pod.

## Proposal
Set `kubernetes-cluster-autoscaler-enabled` adds `cluster-autoscaler.kubernetes.io/safe-to-evict` annotation to Longhorn pods that are not backed by a controller, or with local storage volume mounts. To avoid data loss, Longhorn does not annotate the backing image manager and share manager pods.

Currently, Longhorn creates instance-manager PDBs for replica/engine regardless of the volume state.
During scale down, CA tries to find a removable node but failed by those instance-manager PDBs. We can remove the PDB when the volume is not attached and there is at least one schedulable node with the volume replica and PDB.

### User Stories

#### CA scaling
Before the enhancement, CA will be blocked by
- Pods that are not backed by a controller (engine/replica instance manager).
- Pods with [local storage volume mounts](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/utils/drain/drain.go#L222) (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, csi-snapshotter).

After enhancement, the user can set a new global setting `kubernetes-cluster-autoscaler-enabled` to unblock CA scaling.

> Longhorn considers a node can be removed when conditions are met, else it will block scale down to avoid potential data loss:

Reviewer: How about pods w/ local storage volume mounts mentioned above? the nodes they are running on will never be eligible for cluster autoscaler, right?

Author: Reworded.

- all volumes on the node are detached and there is an available healthy replica on other nodes.
- backing image manager not running on the node.

> - share-manager node not running on the node.

Reviewer: You can reuse the goals you mentioned above because these are the same.

Author: Maybe we don't need to duplicate this, I've removed this block.

### User Experience In Detail

- Configure the setting via Longhorn UI or kubectl.
- Ensure all volume replica count is set to more than 1.
- CA is not blocked by Longhorn components when the node doesn't contain volume replica, backing image manager pod, and share manager pod.

>     - Engine/Replica instance-managerPDB will block the node if the volume is attached.

Author: Fixed.

    - Replica instance-manager PDB will block the node when CA tries to delete the last node with the volume replica.

### API changes

`None`

## Design

### Implementation Overview

#### Global setting
- Add new global setting `Kubernetes Cluster Autoscaler Enabled (Experimental)`.
  - The setting is `boolean`.
  - The default value is `false`.

#### Annotations

> When setting `kubernetes-cluster-autoscaler-enabled` to `true`:

Reviewer 2: ```suggestion
When setting `kubernetes-cluster-autoscaler-enabled` is `true`, Longhorn will add annotation `cluster-autoscaler.kubernetes.io/safe-to-evict` for the following pods:
```

Author: Updated.

> - Annotate engine and replica instance-manager pods when no IM not in state stop/error.

Reviewer 2: I am confused about this. Why cannot we annotate the IM pods when the IM is state stop or error? As long as the PDBs are handled correctly, I don't think we need to consider the pods.

Author: > I am confused about this. Why cannot we annotate the IM pods when the IM is state stop or error? 

We can probably disregard the IM state. Updated.

> As long as the PDBs are handled correctly, I don't think we need to consider the pods.

We need to consider the pods because they are bare pods and also using local storage mounts that will not pass the CA pre-check.

> - Annotate managed deployments with `cluster-autoscaler.kubernetes.io/safe-to-evict`. Managed deployments are labeled with `longhorn.io/managed-by: longhorn-manager` (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, and csi-snapshotter).

Reviewer: ```suggestion
- Annotate managed deployments with local storage volume mounts with `cluster-autoscaler.kubernetes.io/safe-to-evict`. Managed deployments are labeled with `longhorn.io/managed-by: longhorn-manager` (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, and csi-snapshotter).
```

Author: Fixed.

> - Annotate managed deployments using local storage volume mounts with `cluster-autoscaler.kubernetes.io/safe-to-evict`. Managed deployments are labeled with `longhorn.io/managed-by: longhorn-manager` (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, and csi-snapshotter).

Reviewer 2: ```suggestion
The workloads that are managed by longhorn-manager and contain local storage. The managed components are labeled with ......
```
BTW, do we need to handle longhorn-manager?

Author: Updated.

Author: > BTW, do we need to handle longhorn-manager?

No. CA pre-check bypasses DaemonSets.

#### PodDisruptionBudget

- No change to the logic to cleanup PDB if instance-manager doesn't exist.

- Engine IM PDB:
    - Delete PDB if volumes are detached;
        - There is no instance process in IM (im.Status.Instance).
        - The same logic applies when a node is un-schedulable. Node is un-schedulable when marked in spec or with CA tainted `ToBeDeletedByClusterAutoscaler`;
    - Create PDB if volumes are attached; there are instance processes in IM (im.Status.Instance).

- Replica IM PDB:
    - Delete PDB if setting `allow-node-drain-with-last-healthy-replica` is enabled.
    - Delete PDB if volumes are detached;

>         - There is no engine instance process in IM (im.Status.Instance)

Reviewer 2: NIT: When handling the Replica IM PDB, there is no need to consider engine processes.

Author: Fixed.

        - There are other schedulable nodes with healthy volume replica and have replica IM PDB.
    - Delete PDB when a node is un-schedulable. Node is un-schedulable when marked in spec or with CA tainted `ToBeDeletedByClusterAutoscaler`;
        - Check if the condition is met to delete PDB (same check as to when volumes are detached).
        - Enqueue the replica instance-manager of another schedulable node with the volume replica.
        - Delete PBD.
    - Create PDB if volumes are attached:

>         - There are instance processes in IM (im.Status.Instance).

Reviewer 2: If the processes are redundant replica processes, will the replica IM PDB be removed then get recreated immediately?

Reviewer: This seems like a side-effect caused by the redundant replicas, so probably we should not tackle this here?

Reviewer 2: I mean, the replica on the node is not the only healthy one of a volume. In the above deletion logic, the PDB will be removed since this kind of replica won't prevent the deletion. But the controller will recreate the PDB immediately based on this line.

Author: We will only remove the PDB of the detached volume, and only if there is another node with the same volume replica IM PBD. In case all replicas are on the same node of the detached volume, then all replica IM PDB will get created and none will get removed.

Reviewer 2: the PDB of the detached volume?? another node with the same volume replica IM PBD?? Sorry I cannot follow...

Reviewer 2: I think we can describe this part from another perspective: The IM PDB will be created **and retained** only when it's required. And once it's required, we cannot delete the PDB. Required means either of the following happen:
  1. There are volumes/engines running on the node. We need to guarantee that the volumes won't crash
  2. The only available/valid replica of a volume is on the node, no matter whether the volume itself is detached or not. Here we need to prevent the volume data from being lost.

Author: 1. The behavior of creating engine/replica IM PBD  is the same as now when there are volumes attached.
2. We keep only the replica IM PBD of 1 of the nodes when the volume is detached.

> I think we can describe this part from another perspective: The IM PDB will be created **and retained** only when it's required. And once it's required, we cannot delete the PDB. Required means either of the following happen:
> 
> 1. There are volumes/engines running on the node. We need to guarantee that the volumes won't crash
> 2. The only available/valid replica of a volume is on the node, no matter whether the volume itself is detached or not. Here we need to prevent the volume data from being lost.

The implementation overview is written based on scenarios, and I think we are in the same direction.

I've put your suggestion of the description under the proposal section to clarify what we want to achieve.

    - Create PDB when volumes are detached;
        - There is no instance process in IM (im.Status.Instance)
        - The replica has been started. There are no other schedulable nodes with healthy volume replica and have replica IM PDB.

### Test plan

#### Scenario: test CA

    Given Cluster with Kubernetes cluster-autoscaler.
    And Longhorn installed.
    And Set `kubernetes-cluster-autoscaler-enabled` to `true`.

>     And Create deployment with cpu request.

Reviewer: Is the deployment having a volume?

Author: The deployment does not need volume, this is used to trigger CA scaling and will contain CPU requests.
```
        resources:
          limits:
            cpu: 300m
            memory: 30Mi
          requests:
            cpu: 150m
            memory: 15Mi
```

Reviewer 2: You can put this example here.

Author: Updated.

Reviewer 2: Should the CPU request be large enough so that the current nodes cannot fit the requirement?

Author: We can calculate the number of deployment replicas with the allocatable CPU.
```
10 * math.ceil(allocatable_millicpu/cpu_request*node_number/10)
```

Reviewer 2: NIT: The calculation result is for increasing 1-node scaling-up,  right? You can add this expectation there.

Author: The expectation is to double the cluster node number and I have it with the following `When Trigger CA to scale-up by increase deployment replicas.`

>     When Trigger CA to scale-up by increase deployment replicas.

Reviewer: Do you mean the deployment replicas w/ node affinity to enforce a new node is needed?

Author: The deployment does not need node affinity. We will increase the deployment replica numbers and let it consume more CPU resources than the cluster nodes can provide. This triggers CA to scale up and vice versa.

         (double the node number, not including host node)
    Then Cluster should have double the node number.

    When Trigger CA to scale-down by decrease deployment replicas.
         (original node number)
    Then Cluster should have original node number.

#### Scenario: test CA scale down all nodes containing volume replicas

    Given Cluster with Kubernetes cluster-autoscaler.
    And Longhorn installed.
    And Set `kubernetes-cluster-autoscaler-enabled` to `true`.
    And Create volume.
    And Attach the volume.
    And Write some data to volume.
    And Detach the volume.
    And Create deployment with cpu request.

    When Trigger CA to scale-up by increase deployment replicas.
         (double the node number, not including host node)
    Then Cluster should have double the node number.

>     When Annotate new nodes with `scale-down-disable`.

Reviewer: Do you mean `cluster-autoscaler.kubernetes.io/scale-down-disabled`? It's better to describe clearly.

Author: Fixed.

         (this ensures scale-down only the old nodes)
    And Trigger CA to scale-down by decrease deployment replicas.
        (original node number)
    Then Cluster should have original node number + 1 blocked node.

    When Attach the volume to a new node. This triggers replica rebuild.
    And Volume data should be the same.
    And Detach the volume.
    Then Cluster should have original node number.

>     And Volume data should be the same.

Author: Are you referring to `And Volume data should be the same.`?
This is followed by the `Then` statement above.

>     And Volume data should be the same.

Reviewer 2: I think what happens here is:
1. Volume rebuilds new replicas on other nodes.
2. The CA finally evicts the blocking node which previously contains the only replica. This means the cluster node number should be the same as that at the beginning.
3. The volume data is still intact after the rebuilding & eviction.

You may need to explicitly mean the 2nd behavior & result.

> #### Scenario: test CA should block scale down of node running backing image manager pod

Reviewer: ```suggestion
#### Scenario: test CA should block scale down of node running backing image manager/share manager pod
```

Author: https://github.com/longhorn/longhorn/pull/3831#discussion_r847911269

### API changes

`None`

## Design

### Implementation Overview

#### Annotations

When setting `kubernetes-cluster-autoscaler-enabled` to `true`:
- Annotate engine and replica instance-manager pods when no IM not in state stop/error.
- Annotate longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, and csi-snapshotter.

#### PodDisruptionBudget

- Engine IM PDB:
    - Delete PBD if volumes are detached; there is no instance process in IM (im.Status.Instance).
    - Create PBD if volumes are attached; there are instance processes in IM (im.Status.Instance).

- Replica IM PBD:
    - Delete PBD if setting `allow-node-drain-with-last-healthy-replica` is enabled.
    - Delete PBD if volumes are detached;
        - There is instance process in IM (im.Status.Instance)
        - There are other schedulable nodes with healthy volume replica and have replica IM PDB.
    - Create PBD if volumes are attached; there are instance processes in IM (im.Status.Instance).
    - Create PBD when volumes are detached;
        - There is no instance process in IM (im.Status.Instance)

>         - The replica has been started. There are no other schedulable nodes with healthy volume replica and have replica IM PDB.

Reviewer 2: How about other corner cases I mentioned before? Maybe we can revisit this after the implementation

Reviewer 2: I mean, the node is already gone/tainted, or the IM or IM pod is already gone. Some cases are already handled in the current implementation. You may need to reorganize them during this implementation.

Author: Added the implementation to cover node un-schedulable case.

### Test plan

#### Scenario: test CA

Given Cluster of 2 nodes with Kubernetes cluster-autoscaler.
And Longhorn installed.
And Set `kubernetes-cluster-autoscaler-enabled` to `true`.

When create deployment with cpu request.
And trigger CA scale-up 2 nodes by increase deployment replicas.
And trigger CA scale-down 2 nodes by decrease deployment replicas.

Then CA should scale down 1 node.
And cluster should have 3 nodes.
And CA should scale down 1 node.

> And cluster should have 2 nodes.

Reviewer 2: The statement here is a little bit confusing...

Author: Updated.

#### Scenario: test CA scale down all nodes containing volume replicas

Given Cluster of 2 nodes with Kubernetes cluster-autoscaler.
And Longhorn installed.
And Set `kubernetes-cluster-autoscaler-enabled` to `true`.
And Create volume with 2 replicas.
And Attach the volume to start the replicas.
And Detach volume.

When create deployment with cpu request.
And trigger CA scale-up 2 nodes by increase deployment replicas.
And annotate new nodes with `scale-down-disable` to test when CA tries to scale-down all nodes containing volume replica.
And trigger CA scale-down 2 nodes by decrease deployment replicas.

Then CA should scale down 1 node.
And cluster should have 3 nodes.

> And CA scale-down should be blocked by replica instance-manager pod PDB.

Author: I am thinking we can delete the engine IM PBD but keep at least 1 replica IM PBD when the volume is detached.

When create deployment with cpu request.
And trigger CA scale-up 2 nodes by increase deployment replicas.
And annotate new nodes with `scale-down-disable` to test when CA tries to scale-down all nodes containing volume replica.
And trigger CA scale-down 2 nodes by decrease deployment replicas.

Then CA should scale down 1 node.
And cluster should have 3 nodes.
And CA scale-down should be blocked by backing image manager local storage mount.

#### Scenario: test CA should block scale down of node running share manager pod

Given Cluster of 2 nodes with Kubernetes cluster-autoscaler.
And Longhorn installed.
And Set storage class `numberOfReplicas` to `2`.
And Set `kubernetes-cluster-autoscaler-enabled` to `true`.
And Create volume with 'accessMode' rwx.
And Attach the volume to start the replicas.
And Detach volume.

When create deployment with cpu request.
And trigger CA scale-up 2 nodes by increase deployment replicas.
And annotate new nodes with `scale-down-disable` to test when CA tries to scale-down all nodes containing volume replica.
And trigger CA scale-down 2 nodes by decrease deployment replicas.

Then CA should scale down 1 node.
And cluster should have 3 nodes.

> And CA scale-down should be blocked by share manager local storage mount.

Author: Removed.

Set `kubernetes-cluster-autoscaler-enabled` adds `cluster-autoscaler.kubernetes.io/safe-to-evict` annotation to Longhorn pods that are not backed by a controller, or with local storage volume mounts. To avoid data loss, Longhorn does not annotate the backing image manager and share manager pods.

Currently, Longhorn creates instance-manager PDBs for replica/engine regardless of the volume state.
During scale down, CA tries to find a removable node but failed by those instance-manager PDBs.

We can add IM PDB handling to create and retained when the PDB is required:

- There are volumes/engines running on the node. We need to guarantee that the volumes won't crash.
- The only available/valid replica of a volume is on the node. Here we need to prevent the volume data from being lost.

### User Stories

#### CA scaling
Before the enhancement, CA will be blocked by
- Pods that are not backed by a controller (engine/replica instance manager).
- Pods with [local storage volume mounts](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/utils/drain/drain.go#L222) (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, csi-snapshotter).

> After enhancement, instance manager PBD will be actively managed by Longhorn:

Reviewer 2: typo `PBD` -> `PDB`. Same below

#### PodDisruptionBudget

- No change to the logic to cleanup PDB if instance-manager doesn't exist.

- Engine IM PDB:
    - Delete PDB if volumes are detached;
        - There is no instance process in IM (im.Status.Instance).
        - The same logic applies when a node is un-schedulable. Node is un-schedulable when marked in spec or with CA tainted `ToBeDeletedByClusterAutoscaler`;
    - Create PDB if volumes are attached; there are instance processes in IM (im.Status.Instance).

- Replica IM PDB:
    - Delete PDB if setting `allow-node-drain-with-last-healthy-replica` is enabled.
    - Delete PDB if volumes are detached;
        - There is no instance process in IM (im.Status.Instance)
        - There are other schedulable nodes with healthy volume replica and have replica IM PDB.
    - Delete PDB when a node is un-schedulable. Node is un-schedulable when marked in spec or with CA tainted `ToBeDeletedByClusterAutoscaler`;
        - Check if the condition is met to delete PDB (same check as to when volumes are detached).

>         - Enqueue the replica instance-manager of another schedulable node with the volume replica.

Reviewer: Ensure not Enqueue.

Author: We want to enqueue another schedulable replica IM to create the PDB when the current is un-schedulable.

Reviewer: Thanks, it's clear now, but it was not easy to understand a bit.

The point is to ensure at least one healthy replica is still available in another node when the volume is detached and the volume owner node is being eligible for CA by removing it's PDB.

> # Support Kubernetes CA

Author: Addressed.

### Related Issues

https://github.com/longhorn/longhorn/issues/2203

## Motivation

### Goals

- Longhorn should not block CA from scaling.

### Non-goals [optional]

- CA setup.
- CA blocked by kube-system components.

## Proposal
Enable setting `kubernetes-autoscaler-support` adds `cluster-autoscaler.kubernetes.io/safe-to-evict` annotation to all Longhorn pods that are not backed by a controller, or with local storage volume mounts.

Currently, Longhorn creates instance-manager PDBs for replica/engine regardless of the volume state.
During scale down, CA tries to find a removable node but failed by those instance-manager PDBs. We can remove the PDB when the volume is not attached and at least one replica is available.

### User Stories

#### CA scaling
Before the enhancement, CA will be blocked by
- Pods that are not backed by a controller (engine/replica instance manager).

> - Pods with [local storage volume mounts]( local storage volume mounts) (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, csi-snapshotter).

Reviewer 2: How about share manager pods and backing image datasource/manager pods?

For backing image manager pods, if the disk/node contains the only file of one backing image, then you cannot add the annotation for the pod.

For share manager pods, the eviction means you need to stop the RWX volume service for a while and relaunch them on other nodes. Probably we cannot do anything for this until HA RWX support is done.

Author: ATM I don't think there is anything we could do about backing image datasource/manager, and share manager pods. The nodes running with those pods will be blocked by CA local storage condition.

Reviewer 2: For the backing image manager, I think we can do something here. 
As I mentioned, if all backing image files on the node contain duplicates, it's fine to add annotation `cluster-autoscaler.kubernetes.io/safe-to-evict` for the related backing image manager.
The backing image managers won't remove the related files immediately when there is no replica using them. There is a setting named `Backing Image Cleanup Wait Interval` controlling it. Hence, this may unnecessarily block CA.
Of course, the handling of the backing image manager is nice to have. If it makes the logic too complicated, we can skip it now.

Author: After the internal meeting discussion, this feature will be marked experiential and will skip handling of the CA blocked by backing image manager/share manager pod for now.

- Pods with [local storage volume mounts]( local storage volume mounts) (longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, csi-snapshotter).

After enhancement, the user can enable the new global setting `kubernetes-autoscaler-support` to unblock CA scaling. 

Longhorn allows scale down when all volumes on the node are detached and there is an available healthy replica on other nodes. If none of the nodes meets the condition, Longhorn will block scale down with PDB to avoid data loss.

### User Experience In Detail

- Configure the setting via Longhorn UI or kubectl.
- Ensure all volume replica count is set to more than 1.
- PDB will block the replica instance-manager pod when CA tries to delete all nodes with the volume replica.

### API changes

`None`

## Design

### Implementation Overview

#### Annotations

- Annotate engine and replica instance-manager pods when no instance process inside the instance manager.
- Annotate longhorn-ui, longhorn-csi-plugin, csi-attacher, csi-provisioner, csi-resizer, and csi-snapshotter.

#### PodDisruptionBudget

- When the volume is attached, the logic of creating engine/replica IM PBD is the same.

> - Delete engine IM PBD when the volume is detached.

Reviewer 2: Do you mean actively deleting the PDB when there is no process in the IM? It's better to add more details here. e.g., the new PDB sync-up function logic would be like:
1. Check if the PDB can be deleted (I am not sure if the following conditions are corrects or not, will leave them to you for further verifications)
    1. If the node is already gone/tainted/...
    2. If the IM or IM pod is already gone
    3. For engine IM PDB, if there is no engine process......
    4. For replica IM PDB, If there is no replica scheduled......
2. Check if the PDB can be created
    1. volume attachment requirement
    2. replicas are scheduled
    3. ......

- Keep/create one replica IM PBD when the volume is detached.
- When the IM node is unschedulable (tainted by CA), delete the replica IM PBD when there are other schedulable nodes. Before deleting the replica IM PBD, enqueue the schedulable node IM.

> ### Test plan

Reviewer 2: As I mentioned above, need to have tests with backing images or share managers

Author: Added.

> Given Enable `kubernetes-autoscaler-support` setting.

Reviewer 2: How about the cluster config?

Author: Updated.

And Create volume with 2 replicas.
And Attach the volume to start the replicas.
And Detach volume.

When create deployment with cpu request.
And trigger CA scale-up 2 nodes by increate deployment replicas.

> And annotate new nodes with `scale-down-disable` to test when CA tries to scale-down all volume replica nodes.

Reviewer 2: You mean disabling the scaling down for the new node that does not contain any volume replica, then check if CA could kick out the 2 other nodes?

Author: Reworded.
